test/functional/km/control.cpp

```diff
@@ -0,0 +1,489 @@
+INITCODE
+NTSTATUS
+TestDrvCtlInitialize(
+    _In_ WDFDRIVER Driver
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+    PWDFDEVICE_INIT DeviceInit = nullptr;
+    WDF_FILEOBJECT_CONFIG FileConfig;
+    WDF_OBJECT_ATTRIBUTES Attribs;
+    WDFDEVICE Device;
+    WDF_IO_QUEUE_CONFIG QueueConfig;
+    WDFQUEUE Queue;
+
+    DeviceInit =
+        WdfControlDeviceInitAllocate(
+            Driver,
+            &SDDL_DEVOBJ_SYS_ALL_ADM_ALL);
+    if (DeviceInit == nullptr) {
+        TraceError(
+            "[ lib] ERROR, %s.",
+            "WdfControlDeviceInitAllocate failed");
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto Error;
+    }
+
+    Status =
+        WdfDeviceInitAssignName(
+            DeviceInit,
+            &TestDrvCtlDeviceName);
+    if (!NT_SUCCESS(Status)) {
+        TraceError(
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "WdfDeviceInitAssignName failed");
+        goto Error;
+    }
+
+    WDF_FILEOBJECT_CONFIG_INIT(
+        &FileConfig,
+        TestDrvCtlEvtFileCreate,
+        TestDrvCtlEvtFileClose,
+        TestDrvCtlEvtFileCleanup);
+    FileConfig.FileObjectClass = WdfFileObjectWdfCanUseFsContext2;
+
+    WDF_OBJECT_ATTRIBUTES_INIT_CONTEXT_TYPE(&Attribs, TEST_CLIENT);
+    WdfDeviceInitSetFileObjectConfig(
+        DeviceInit,
+        &FileConfig,
+        &Attribs);
+
+    Status =
+        WdfDeviceCreate(
+            &DeviceInit,
+            &Attribs,
+            &Device);
+    if (!NT_SUCCESS(Status)) {
+        TraceError(
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "WdfDeviceCreate failed");
+        goto Error;
+    }
+
+    Status = WdfDeviceCreateSymbolicLink(Device, &TestDrvCtlDeviceSymLink);
+    if (!NT_SUCCESS(Status)) {
+        TraceError(
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "WdfDeviceCreateSymbolicLink failed");
+        goto Error;
+    }
+
+    WDF_IO_QUEUE_CONFIG_INIT_DEFAULT_QUEUE(&QueueConfig, WdfIoQueueDispatchParallel);
+    QueueConfig.EvtIoDeviceControl = TestDrvCtlEvtIoDeviceControl;
+    QueueConfig.EvtIoCanceledOnQueue = TestDrvCtlEvtIoCanceled;
+
+    __analysis_assume(QueueConfig.EvtIoStop != 0);
+    Status =
+        WdfIoQueueCreate(
+            Device,
+            &QueueConfig,
+            WDF_NO_OBJECT_ATTRIBUTES,
+            &Queue);
+    __analysis_assume(QueueConfig.EvtIoStop == 0);
+
+    if (!NT_SUCCESS(Status)) {
+        TraceError(
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "WdfIoQueueCreate failed");
+        goto Error;
+    }
+
+    TestDrvCtlDevice = Device;
+
+    WdfControlFinishInitializing(Device);
+
+    TraceVerbose(
+        "[test] Control interface initialized");
+
+Error:
+
+    if (DeviceInit) {
+        WdfDeviceInitFree(DeviceInit);
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+VOID
+TestDrvCtlUninitialize(
+    )
+{
+    TraceVerbose(
+        "[test] Control interface uninitializing");
```

Comment:
why is this `[test]` and above and below are `[ lib]`?

Reply:
@nigriMSFT ANy idea?

Reply:
As mentioned in another comment, lots of this was copied from msquic. Probably mixed with some new bits. Please choose a consistent pattern that makes sense in this repo